### PR TITLE
Add concept exercise: dicts / high-score

### PR DESCRIPTION
**Pull request by SaschaMann**

### CODE REVIEW

File: exercises/concept/fibonacci-iterator/.docs/introduction.md [hunk 1]
```diff
@@ -85,7 +85,11 @@ Base.iterate(S::Squares) = iterate(S, 1)
 
 You may find it useful to combine these two definitions into one using optional arguments:
 
-## optional-arguments
+## Making use of optional arguments
```
Comment (SaschaMann) [inline exercises/concept/fibonacci-iterator/.docs/introduction.md:88]: I feel like it makes sense to quickly go over optional arguments in the introduction for `fibonacci-iterator` again, even if it's technically already introduced in the new `high-score` exercise.

By the time the student gets to this, they may have forgotten and it's commonly used for defining iterators.

File: exercises/concept/high-score/.docs/instructions.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+In this exercise, you're implementing a way to keep track of the high scores for the most popular game in your local arcade hall.
+
+<!-- TODO
+    Decide if "edge cases" should be listed in the instructions or not.
+
+    By explicitely mentioning them, students may consider them ahead of time when they don't really need to, because the edge cases match the behaviour of the naive implementation anyway.
+
+    However, if students deviate from the naive implementation, they may find it frustrating that tests fail even if their implementation conforms to the instructions.
+-->
```
Comment (SaschaMann) [inline exercises/concept/high-score/.docs/instructions.md:11]: This refers to edge cases such as "What happens if the score of a player that isn't on the board is reset?".

I'd prefer not mentioning them in the instructions for the reasons in the comment but I don't have a strong opinion on it and am happy to change it if you think it's better to cover every case in the instructions.

File: exercises/concept/high-score/.meta/design.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+## Considered but dropped
+
+Implement a `high_score` function that returns the leader, e.g.
+
+```julia
+function high_score(scores)
+    best = findmax(scores)
+    best[2] => best[1]
+end
+```
```
Comment (SaschaMann) [inline exercises/concept/high-score/.meta/design.md:12]: This doesn't actually work because it doesn't handle the edge cases, I merely left it in as an example for future reference.

**Comment by SaschaMann** [comment]
One thing I'm kinda missing is a task about _defining_ a dictionary rather than manipulating it. That said, I assume for testing they'll like do so in the REPL and since there's not much to it, I think it's fine to not have an explicit task about it.

That said, do you have any ideas for a task that requires explicitely defining a Dict?

File: exercises/concept/high-score/.docs/introduction.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Introduction
+
+<!-- TODO: As you may recall from Elyse's enchantments, arrays also use push!() -->
```
Comment (cmcaine) [inline exercises/concept/high-score/.docs/introduction.md:3]: Where will students use `push!` in this exercise?

File: exercises/concept/high-score/.meta/exemplar.jl [hunk 5]
```diff
@@ -0,0 +1,34 @@
+"""
+    add_player!(scores, name, score=0)
+
+Add a player to the score board.
+"""
+add_player!(scores, name, score=0) = push!(scores, name => score)
```
Comment (cmcaine) [inline exercises/concept/high-score/.meta/exemplar.jl:6]: No action required, but your earlier comment about `push!` confused me because I always do this as `scores[name] = score`. Didn't even know this was possible!

File: exercises/concept/high-score/.meta/exemplar.jl [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+Update the score of a player on the score board.
+"""
+update_score!(scores, name, score) = scores[name] = score
+
+"""
+    reset_score!(scores, name)
+
+Reset the score of a player on the score board.
+"""
+reset_score!(scores, name) = update_score!(scores, name, 0)
```
Comment (cmcaine) [inline exercises/concept/high-score/.meta/exemplar.jl:27]: Can do three of these at once with this slightly confusing line ;)

```julia
const add_player! = reset_score! = update_score!(scores, name, score=0) = scores[name] = score
```

File: exercises/concept/high-score/.docs/instructions.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+In this exercise, you're implementing a way to keep track of the high scores for the most popular game in your local arcade hall.
+
+## 1. Add players to the high score board
+
+To add a player to the high score board, define two `add_player!` methods that take the following arguments:
+
+- The first argument is the dictionary of scores.
+- The second argument is the name of a player as a string.
+- The third argument is the score as an integer.
+  The argument is optional, when no value is given, the score should be set to 0.
```
Comment (cmcaine) [inline exercises/concept/high-score/.docs/instructions.md:12]:
```suggestion
  This argument is optional, when no value is given, the score should be set to 0.
```

**Review by cmcaine — APPROVED** [review]
Looks good to me! I think it's a smart move to avoid specifying the return types.

File: exercises/concept/high-score/.meta/exemplar.jl (hunk 6 above)
Comment (SaschaMann) [reply exercises/concept/high-score/.meta/exemplar.jl:27]: That's pretty neat :D

**Comment by SaschaMann** [comment]
Thanks for the review! :)

File: exercises/concept/high-score/.meta/exemplar.jl (hunk 5 above)
Comment (SaschaMann) [reply exercises/concept/high-score/.meta/exemplar.jl:6]: I'd usually also do it as `scores[name] = score` but I think there's some benefit to introduce people to the idea that `push!` works on a large number of collections, even if in this particular case it may not be the most common use. We still show off `scores[name] = score` in the update method below.

Technically `push!(scores, name => score)` can even _update_ the value of `scores[name]` if `name` is already in `scores` but I find that use of `push!` slightly misleading.
